monai/data/wsi_datasets.py

```diff
@@ -0,0 +1,116 @@
+
+from monai.data import Dataset
+from monai.data.wsi_reader import WSIReader
+from monai.utils import ensure_tuple_rep
+
+__all__ = ["PatchWSIDataset"]
+
+
+class PatchWSIDataset(Dataset):
+    """
+    This dataset extracts patches from whole slide images (without loading the whole image)
+    It also reads labels for each patch and provides each patch with its associated class labels.
+
+    Args:
+        data: the list of input samples including image, location, and label (see the note below for more details).
+        size: the size of patch to be extracted from the whole slide image.
+        level: the level at which the patches to be extracted (default to 0).
+        transform: transforms to be executed on input data.
+        reader_name: the name of library to be used for loading whole slide imaging, as the backend of `monai.data.WSIReader`
+            Defaults to CuCIM.
+
+    Note:
+        The input data has the following form as an example:
+
+        .. code-block:: python
+
+            [
+                {"image": "path/to/image1.tiff", "location": [200, 500], "label": 0},
+                {"image": "path/to/image2.tiff", "location": [100, 700], "label": 1}
+            ]
+
+    """
+
+    def __init__(
+        self,
+        data: List,
+        size: Optional[Union[int, Tuple[int, int]]] = None,
+        level: Optional[int] = None,
+        transform: Optional[Callable] = None,
+        reader_name: str = "cuCIM",
+    ):
+        super().__init__(data, transform)
+
+        # Ensure patch size is a two dimensional tuple
+        if size is None:
+            self.size = None
+        else:
+            self.size = ensure_tuple_rep(size, 2)
+
+        # Create a default level
+        self.level = level
+
+        # Setup the WSI reader backend
+        self.reader_name = reader_name.lower()
+        self.image_reader = WSIReader(backend=self.reader_name)
+
+        # Initialized an empty whole slide image object dict
+        self.wsi_object_dict: Dict = {}
+
+    def _get_wsi_object(self, sample):
+        image_path = sample["image"]
+        if image_path not in self.wsi_object_dict:
+            self.wsi_object_dict[image_path] = self.image_reader.read(image_path)
+        return self.wsi_object_dict[image_path]
+
+    def _empty_wsi_object_dict(self):
+        self.wsi_object_dict = {}
+
+    def _get_label(self, sample):
+        return np.array(sample["label"], dtype=np.float32)
+
+    def _get_location(self, sample):
+        size = self._get_size(sample)
+        return [sample["location"][i] - size[i] // 2 for i in range(len(size))]
+
+    def _get_level(self, sample):
+        if self.level is None:
+            return sample.get("level")
+        return self.level
+
+    def _get_size(self, sample):
+        if self.size is None:
+            return ensure_tuple_rep(sample.get("size"), 2)
+        return self.size
+
+    def _get_data(self, sample):
+        if self.reader_name == "openslide":
```

Comment:
I don't quite understand why only empty the objects dictionary when reader is `openslide`?

Thanks.

Reply:
`OpenSlide` has a caching mechanism that we want to avoid here since we don't want to cache the tiles that related to the already extracted patch. It can create problems when extracting thousands of patches from hundreds of whole slide images.

Reply:
OK, I see. Thanks for your explanation.
Could you please help add some comments or doc-string to describe it, to help others understand the code?

Thanks in advance.

Reply:
Sure